### PR TITLE
[WIP] add "how to boot from USB stick" info to installer

### PR DESCRIPTION
**Pull request by deeplow**
Fixes QubesOS/qubes-issues/issues/6758

### CODE REVIEW

**Comment by andrewdavidwong** [comment]
Regarding the use of `<kbd>`:

Although the HTML tags are technically correct, it's probably more
important to preserve source readability and avoid the problems that
HTML-in-Markdown poses for localization and offline documentation.

File: user/downloading-installing-upgrading/installation-guide.md
Comment (andrewdavidwong) [inline user/downloading-installing-upgrading/installation-guide.md:158]: I think it might look somewhat strange to have two info alerts back-to-back like this. What do you think?

File: user/downloading-installing-upgrading/installation-guide.md
Comment (andrewdavidwong) [inline user/downloading-installing-upgrading/installation-guide.md:139]: What about users who are using (U)EFI instead of BIOS?

File: user/downloading-installing-upgrading/installation-guide.md
Comment (andrewdavidwong) [inline user/downloading-installing-upgrading/installation-guide.md:149]: You can put this copyright information in [COPYRIGHT.md](https://github.com/QubesOS/qubes-attachment/blob/master/COPYRIGHT.md) instead.

Also, it's good practice to include something in the title rather than leaving it blank (i.e., between the square brackets in the `![]` part).

I'm also not sure why we're using a table here and what "ADD here UEFI" means. If that's supposed to be a placeholder for a future image, maybe make it an HTML comment or something.

Also, don't forget to submit a PR against https://github.com/QubesOS/qubes-attachment/tree/master with your image(s).

File: user/downloading-installing-upgrading/installation-guide.md
Comment (andrewdavidwong) [inline user/downloading-installing-upgrading/installation-guide.md:145]: Note that this say "BIOS" but then also mentions a UEFI image.

File: user/downloading-installing-upgrading/installation-guide.md
Comment (andrewdavidwong) [inline user/downloading-installing-upgrading/installation-guide.md:157]: These asterisks will be literal asterisks, not italics, since you're currently inside of an alert div. You'll have to use `<em>` here if you want emphasis, but I don't think emphasis makes anything clearer in this case.

Note that there are some typos and unclear language here. I can do a copyediting pass to fix these before merging, but in case you're curious what I mean, here are some examples:
- "other thing" -> "other things"
- "in BIOS" -> "in the BIOS"
- "You may need to change them later" -- Change what?
- "deactivating secure boot can solve it" -- I don't think this will solve ever case of the installer refusing to start.
- "if when installing Qubes shows an error" -- I read this as "if, when installing Qubes, shows an error" which makes no sense, but I'm guessing you meant "if, when installing, Qubes shows an error." IMO, punctuation makes a big difference here. But it's also a bit lax to say "Qubes" shows an error. More precise would be "the installer," since Qubes OS isn't even installed yet.

**Review by andrewdavidwong — CHANGES_REQUESTED** [review]

File: user/downloading-installing-upgrading/installation-guide.md
Comment (deeplow) [reply user/downloading-installing-upgrading/installation-guide.md:139]: In my UEFI BIOS it's pretty much the same. Do you have a different experience? Perhaps I should look at more distro's advice

File: user/downloading-installing-upgrading/installation-guide.md
Comment (deeplow) [reply user/downloading-installing-upgrading/installation-guide.md:158]: I was also thinking that. But since github doesn't allow me to draft up a thing before opening a PR it went as it was. I'll deal with that in a future commit.

File: user/downloading-installing-upgrading/installation-guide.md
Comment (deeplow) [reply user/downloading-installing-upgrading/installation-guide.md:149]: Didn't know about COPYRIGHT.md. Will put it there.

"ADD here UEFI" --> because of github's non-existant drafting UI.

I'll also fix the table to make the title "Legacy BIOS" and "UEFI" and will add the image on qubes-attachment.

File: user/downloading-installing-upgrading/installation-guide.md
Comment (deeplow) [reply user/downloading-installing-upgrading/installation-guide.md:157]: I'll revise this. The story is the same. I started working on this and ended up not having enough time. I really wish there was a personal draft thing. Would have saved you a bit of time revising this lower than acceptable WIP. Sorry

Comment (andrewdavidwong) [reply user/downloading-installing-upgrading/installation-guide.md:157]: > I'll revise this. The story is the same. I started working on this and ended up not having enough time. I really wish there was a personal draft thing. Would have saved you a bit of time revising this lower than acceptable WIP. Sorry

Oh, no worries! There actually is a way to make a draft PR. I'm not sure if it requires special permission or something, but it shouldn't:

https://github.blog/2019-02-14-introducing-draft-pull-requests/

File: user/downloading-installing-upgrading/installation-guide.md
Comment (andrewdavidwong) [reply user/downloading-installing-upgrading/installation-guide.md:139]: > In my UEFI BIOS it's pretty much the same. Do you have a different experience? Perhaps I should look at more distro's advice

Well, it's more about the language. Even if the experience is the same, the reader won't know that unless we tell them. If we just say "BIOS," then someone using UEFI may think, "Aw, crap! My computer uses UEFI, not BIOS, so this whole guide doesn't apply to me! :frowning_face:" So it's better to say "'BIOS' or 'UEFI'" or somehow indicate that we're talking about both.

File: user/downloading-installing-upgrading/installation-guide.md
Comment (deeplow) [reply user/downloading-installing-upgrading/installation-guide.md:157]: Yes, I could have made those. But I believe that's just a costmetic thing (very much like "[WIP]" in the title). In realility is is publicly listed just like all other PRs.

Comment (andrewdavidwong) [reply user/downloading-installing-upgrading/installation-guide.md:157]: > Yes, I could have made those. But I believe that's just a costmetic thing (very much like "[WIP]" in the title). In realility is is publicly listed just like all other PRs.

Ah, I actually didn't even realize, until this very moment, that the `[WIP]` in your PR's title meant you didn't want it reviewed yet, lol. Sorry about that. After seeing people include all sorts of `[tags]` in the titles of their issues and PRs over the years (many of them weird and pointless, unlike yours), my brain just filters them out now.

Submitting the PR as a draft, although public, is much better, because I can't merge it while it's a draft, so it makes it absolutely clear what the intention is.

**Comment by deeplow** [comment]
"Replace all BIOS mentions with BIOS/UEFI" (https://github.com/QubesOS/qubes-doc/pull/1172/commits/9d2bc4341f662ed76ecfb6b4ce9b0708cc19d181) -- this may cause more issues than it solves, although technically more correct. Should I undo it?

**Comment by andrewdavidwong** [comment]
> "Replace all BIOS mentions with BIOS/UEFI" (https://github.com/QubesOS/qubes-doc/pull/1172/commits/9d2bc4341f662ed76ecfb6b4ce9b0708cc19d181) -- this may cause more issues than it solves, although technically more correct. Should I undo it?

What do you mean? Like what kinds of issues?

**Comment by deeplow** [comment]
> > "Replace all BIOS mentions with BIOS/UEFI" ([9d2bc43](https://github.com/QubesOS/qubes-doc/commit/9d2bc4341f662ed76ecfb6b4ce9b0708cc19d181)) -- this may cause more issues than it solves, although technically more correct. Should I undo it?
> 
> What do you mean? Like what kinds of issues?

Too much technical jargon. I think to most non-technical users if we referred to UEFI as BIOS it would not make a difference. They would just think "ah, that weird-looking menu". For technical users we could just say for simplification sake whenever we say BIOS we also mean UEFI.

Is there any name for the generalization of these two?

**Comment by andrewdavidwong** [comment]
> > > "Replace all BIOS mentions with BIOS/UEFI" ([9d2bc43](https://github.com/QubesOS/qubes-doc/commit/9d2bc4341f662ed76ecfb6b4ce9b0708cc19d181)) -- this may cause more issues than it solves, although technically more correct. Should I undo it?
> > 
> > 
> > What do you mean? Like what kinds of issues?
> 
> Too much technical jargon. I think to most non-technical users if we referred to UEFI as BIOS it would not make a difference. They would just think "ah, that weird-looking menu". For technical users we could just say for simplification sake whenever we say BIOS we also mean UEFI.
> 

Nontechnical users won't understand either one. Users who understand the difference may be thrown off if we say only "BIOS" (see my comment above).

> Is there any name for the generalization of these two?

Not that I'm aware of. :\

**Comment by deeplow** [comment]
> > 
> 
> Nontechnical users won't understand either one. Users who understand the difference may be thrown off if we say only "BIOS" (see my comment above).
> 
> > Is there any name for the generalization of these two?
> 
> Not that I'm aware of. :\

Ok. So let's maybe keep the consistency of referring to it as BIOS/UEFI only distinguishing them on the images.

**Comment by deeplow** [comment]
I've posted on the forum to ask if someone has feedback on this. https://forum.qubes-os.org/t/feedback-welcome-documenting-how-to-boot-qubes-from-usb-drive/4928

If no feedback is received in maybe two days then I guess it's ready for the final review.

File: user/downloading-installing-upgrading/installation-guide.md
Comment (deeplow) [reply user/downloading-installing-upgrading/installation-guide.md]: removed this entirely.

**Comment by deeplow** [comment]
@andrewdavidwong ready for final review.

**Comment by andrewdavidwong** [comment]
I made some changes: https://github.com/QubesOS/qubes-doc/pull/1172/commits/380e7b43b9494bd603a17f142d6b27dbf70dd8c8.

I actually don't think this belongs in the core Qubes documentation for the reasons explained [here](https://www.qubes-os.org/doc/documentation-style-guide/#core-vs-external-documentation), but I'm willing to give it a try.

**Review by andrewdavidwong — APPROVED** [review]

**Comment by deeplow** [comment]
> I actually don't think this belongs in the core Qubes documentation for the reasons explained [here](https://www.qubes-os.org/doc/documentation-style-guide/#core-vs-external-documentation), but I'm willing to give it a try.

Regardless of whatever rules the core-vs-external has this has to break those rules. Choosing the boot device is an integral part of the installation. We don't want to make users have to go through even more external references about for the "basic" thing that is installing the OS itself. The lower the barrier there (that's very high atm), the less people that are able to actually use it.

**Comment by deeplow** [comment]
Not sure if there is a particular reason for you syncing changes with master this ways. But in case you're not aware already here's a tip: in order to update this branch with the current content of master you can do a [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) instead of a merge. It won't have all of these unrelated commits here. The history will be as clean as if I had originally created it on top of the latest version of the docs.

**Comment by unman** [comment]
On Sun, Jul 11, 2021 at 05:17:32AM -0700, deeplow wrote:
> Not sure if there is a particular reason for you syncing changes with master this ways. But in case you're not aware already here's a tip: in order to update this branch with the current content of master you can do a [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) instead of a merge. It won't have all of these unrelated commits here. The history will be as clean as if I had originally created it on top of the latest version of the docs.
> 

I suspect this is a function of using the GitHub interface - but maybe
wrong.
In any case, not sure what "unrelated commits" you mean. The commits
reflect the development of your final commit, and match the discussion,
don't they? I don't think this is a bad thing.
Perhaps I've missed your point.

**Comment by andrewdavidwong** [comment]
> > I actually don't think this belongs in the core Qubes documentation for the reasons explained [here](https://www.qubes-os.org/doc/documentation-style-guide/#core-vs-external-documentation), but I'm willing to give it a try.
> 
> Regardless of whatever rules the core-vs-external has this has to break those rules. Choosing the boot device is an integral part of the installation. We don't want to make users have to go through even more external references about for the "basic" thing that is installing the OS itself. The lower the barrier there (that's very high atm), the less people that are able to actually use it.

If done correctly, it needn't increase the difficulty in any meaningful way. For example, we could say something like:

> Next, you'll have to set your computer to boot from the USB drive to which you've copied the Qubes installer ISO. If you're not sure how to do that, simply [follow this easy guide](https://www.howtogeek.com/129815/beginner-geek-how-to-change-the-boot-order-in-your-computers-bios/). Go ahead and follow the guide right now. We'll wait.
> 
> Got it? Good. Now that you've set your computer to boot from your USB drive, simply restart your computer and, in less than a minute, you should be greeted with this installer screen...

I'm not saying that this doesn't increase the difficulty *at all*. Rather, my point is that this increase in difficulty is acceptable and not significant relative to what's demanded of users *after* they successfully install Qubes and try to use it. Let's face it: There are a thousand things in Qubes OS (in its current state) that are much more difficult than booting from a USB drive (or, for that matter, following instructions on an external website then resuming where you left off on our website). It almost feels like a bait-and-switch to make it orders of magnitude easier to *install* Qubes than to *use* Qubes. If we hold the user's hand too much and except zero familiarity with technology in the onboarding process, it might seem a bit as if we're leading lambs to the slaughter since we know what awaits them once they actually start using it. Know what I mean?

**Comment by andrewdavidwong** [comment]
> Not sure if there is a particular reason for you syncing changes with master this ways. But in case you're not aware already here's a tip: in order to update this branch with the current content of master you can do a [rebase](https://git-scm.com/book/en/v2/Git-Branching-Rebasing) instead of a merge. It won't have all of these unrelated commits here. The history will be as clean as if I had originally created it on top of the latest version of the docs.

Good tip, thanks! Will try this next time.

> I suspect this is a function of using the GitHub interface - but maybe wrong.

Partially, yes. Here's an example of the commands GitHub recommends to me when I go to merge a PR:

```
Step 1: From your project repository, check out a new branch and test the changes.

git checkout -b deeplow-patch-23 master
git pull https://github.com/deeplow/qubes-doc.git patch-23

Step 2: Merge the changes and update on GitHub.

git checkout master
git merge --no-ff deeplow-patch-23
git push origin master
```

But I'm not forced to follow these commands, so the other part is just my ignorance about Git.

> In any case, not sure what "unrelated commits" you mean. The commits reflect the development of your final commit, and match the discussion, don't they? I don't think this is a bad thing. Perhaps I've missed your point.

He's referring to all of the qubes-doc commits in the `master` branch between the time he opened this PR and my last push to this PR showing up here. There have been a lot of commits from other work unrelated to this PR, and they're all visible here on the web interface. This occurs when I push to a PR after following the "Step 1" instructions above, presumably because my `master` is ahead, and `git pull` does `git fetch` followed by `git merge`. I've previously noted this unintended effect in other PRs where the same thing has occurred and asked whether I was doing things correctly and whether there was a better way. So far, no one has answered that question until @deeplow did just now, so thanks @deeplow! :slightly_smiling_face: 

Strictly speaking, I could avoid this problem by just not pushing commits to PRs at all. I generally do so only as a courtesy to the person who opened the PR so that they can more easily see changes I've made that are directly related to that PR in the course of merging it. They could find the same thing by looking at the main git log to see subsequent commits within the branch before it was merged, or, if I mentioned the PR number in my commit messages, see them linked here. However, people often seem confused when I don't push to the PR. They forget to check the history of `master` to see if any subsequent changes have been made and instead comment on the now-outdated PR. Pushing to the PR seems like good practice, so that's why I do it (but I could be wrong about that). More rarely, collaboration sometimes requires it, so it's good to have a solution either way.

I'm still not entirely clear on what exactly to do differently now, however. Should it be something like this instead of "Step 1"?

```
git checkout -b deeplow-patch-23 master
git fetch https://github.com/deeplow/qubes-doc.git patch-23
git rebase origin master
```

...then as usual after that?

**Comment by deeplow** [comment]
>  In any case, not sure what "unrelated commits" you mean.

I don't think think it's GitHub related. For example the commit https://github.com/QubesOS/qubes-doc/pull/1172/commits/321ee9dedcf3d1eb65f0fd30e1429693bb492a5e show in this branch. Anyways, it's just a minor thing.

**Comment by deeplow** [comment]
> > 
> 
> If done correctly, it needn't increase the difficulty in any meaningful way. For example, we could say something like:
> 
> > Next, you'll have to set your computer to boot from the USB drive to which you've copied the Qubes installer ISO. If you're not sure how to do that, simply [follow this easy guide](https://www.howtogeek.com/129815/beginner-geek-how-to-change-the-boot-order-in-your-computers-bios/). Go ahead and follow the guide right now. We'll wait.
> > Got it? Good. Now that you've set your computer to boot from your USB drive, simply restart your computer and, in less than a minute, you should be greeted with this installer screen...

Oh. Sure. If you find a good guide, just replace all my text with a link to it.

**Comment by deeplow** [comment]
> I'm still not entirely clear on what exactly to do differently now, however. Should it be something like this instead of "Step 1"?
> 
> ```
> git checkout -b deeplow-patch-23 master
> git fetch https://github.com/deeplow/qubes-doc.git patch-23
> git rebase origin master
> ```

Not sure if the middle command is needed. But yes, something like that :)

**Comment by andrewdavidwong** [comment]
> > > 
> > 
> > 
> > If done correctly, it needn't increase the difficulty in any meaningful way. For example, we could say something like:
> > > Next, you'll have to set your computer to boot from the USB drive to which you've copied the Qubes installer ISO. If you're not sure how to do that, simply [follow this easy guide](https://www.howtogeek.com/129815/beginner-geek-how-to-change-the-boot-order-in-your-computers-bios/). Go ahead and follow the guide right now. We'll wait.
> > > Got it? Good. Now that you've set your computer to boot from your USB drive, simply restart your computer and, in less than a minute, you should be greeted with this installer screen...
> 
> Oh. Sure. If you find a good guide, just replace all my text with a link to it.

Well, like I said, I'm willing to give this a try, especially now that it's already merged. It's a relatively small amount of additional content that shouldn't need to change very often. However, if it proves to be too much of a maintenance burden in the future, we can always make that replacement. (For example, if some new thing besides BIOS and UEFI becomes prevalent, and it would be too labor-intensive or time-consuming for us to update this, then that might prompt us to make such a replacement.)

**Comment by andrewdavidwong** [comment]
> > I'm still not entirely clear on what exactly to do differently now, however. Should it be something like this instead of "Step 1"?
> > ```
> > git checkout -b deeplow-patch-23 master
> > git fetch https://github.com/deeplow/qubes-doc.git patch-23
> > git rebase origin master
> > ```
> 
> Not sure if the middle command is needed. But yes, something like that :)

Well, without the middle command, how do I get your changes from GitHub? All I've done in the first command is make a new branch called `deeplow-patch-23` with the exact same content as my local `master` branch, right?

**Comment by deeplow** [comment]
> Well, without the middle command, how do I get your changes from GitHub? All I've done in the first command is make a new branch called `deeplow-patch-23` with the exact same content as my local `master` branch, right?

Ah, you're right. I was thinking more about collaborating in the same repos. I don't have much experience about rebasing from different repos. That may be the way to do it. I think you have a typo and missed `deeplow-` in the second line. I believe the full thing will be:

```
git checkout -b deeplow-patch-23 master
git fetch https://github.com/deeplow/qubes-doc.git deeplow-patch-23
git rebase origin master
git checkout master
git merge deeplow-patch-23
```

First you rebase from master (literally reproducing all the latest commits on top of master), and then once it's up to date you can do a normal merge.

But again, I'm not sure about how doing this across different repos changes the thing. Because you can commit into by branch and that's not normal in git flow, I think.

**Comment by andrewdavidwong** [comment]
> > Well, without the middle command, how do I get your changes from GitHub? All I've done in the first command is make a new branch called `deeplow-patch-23` with the exact same content as my local `master` branch, right?
> 
> Ah, you're right. I was thinking more about collaborating in the same repos. I don't have much experience about rebasing from different repos. That may be the way to do it.

Yeah, almost all PRs are cross-repo because contributors don't have push access to the official repos.

> I think you have a typo and missed `deeplow-` in the second line.

I'm pretty sure it's supposed to be left out, because it's the branch in _your_ repo.

> I believe the full thing will be:
> 
> ```
> git checkout -b deeplow-patch-23 master
> git fetch https://github.com/deeplow/qubes-doc.git deeplow-patch-23
> git rebase origin master
> git checkout master
> git merge deeplow-patch-23
> ```
> 
> First you rebase from master (literally reproducing all the latest commits on top of master), and then once it's up to date you can do a normal merge.
> 
> But again, I'm not sure about how doing this across different repos changes the thing. Because you can commit into by branch and that's not normal in git flow, I think.

Me being able to push to your PR branch is a special thing (which you as the PR owner can choose to allow or deny), but it's also the only reason the original problem can possibly occur in the first place.

**Comment by andrewdavidwong** [comment]
In any case, thanks for the help! Will try it out next time.

**Comment by andrewdavidwong** [comment]
I'm trying this out on a separate PR (https://github.com/QubesOS/qubes-doc/pull/1031), but nothing is happening at this step:

```
$ git rebase origin master
Current branch master is up to date.
```

I know for a fact that there have been a *lot* of commits that are missing from this PR's branch.

I think something has gone wrong before this step, because the log in the PR branch already shows all the latest commits, probably because I created this branch off of the up-to-date master branch. It seems like what I really need is the ability to just replicate the PR's branch directly on my local machine, independently of the state of my own repo.

**Comment by andrewdavidwong** [comment]
I guess I could find the latest commit ID common to both the PR branch and my master branch, then squash all subsequent commits together, then push that to the PR branch, but that seems like it can't be anyone's intended workflow. There must be an easier way.

**Comment by andrewdavidwong** [comment]
I just tried pushing only a single commit by following [this](https://stackoverflow.com/a/21583220), but it still pushed the entire history.

**Comment by andrewdavidwong** [comment]
@fepitre suggested this:

https://stackoverflow.com/questions/27567846/how-can-i-check-out-a-github-pull-request-with-git#30584951

Looks promising. Will try it next time this arises.

**Comment by marmarek** [comment]
> @fepitre suggested this:
> 
> https://stackoverflow.com/questions/27567846/how-can-i-check-out-a-github-pull-request-with-git#30584951
> 
> Looks promising. Will try it next time this arises.

Yes, this should work. I'm not sure if pushing changes to the original branch won't require some extra steps - usually I need to add "remote" pointing at contributor's repository, as those magic "pull/..." branches are read-only on github (but maybe that changed?).

**Pull request merged**
